<a href="https://colab.research.google.com/github/Shambhuraje1919/Machine_Learning_Paractice/blob/main/FeatureEng.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("arnabchaki/data-science-salaries-2023")

print("Path to dataset files:", path)

100%|██████████| 25.4k/25.4k [00:00<00:00, 19.2MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/arnabchaki/data-science-salaries-2023/versions/1


In [4]:
import pandas as pd
import os

In [133]:
os.listdir(path)

['ds_salaries.csv']

In [134]:
df = pd.read_csv(os.path.join(path,"ds_salaries.csv"))

In [89]:
df.sample(5
        )

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
454,2023,MI,FT,Data Analyst,45000,GBP,54685,GB,0,GB,M
6,2023,SE,FT,Applied Scientist,136000,USD,136000,US,0,US,L
1709,2023,SE,FT,Machine Learning Scientist,165750,USD,165750,US,100,US,M
2506,2022,SE,FT,Data Engineer,155000,USD,155000,US,0,US,M
1428,2023,EX,FT,Data Engineer,239000,USD,239000,US,100,US,M


In [90]:
df["job_title"].unique()

array(['Principal Data Scientist', 'ML Engineer', 'Data Scientist',
       'Applied Scientist', 'Data Analyst', 'Data Modeler',
       'Research Engineer', 'Analytics Engineer',
       'Business Intelligence Engineer', 'Machine Learning Engineer',
       'Data Strategist', 'Data Engineer', 'Computer Vision Engineer',
       'Data Quality Analyst', 'Compliance Data Analyst',
       'Data Architect', 'Applied Machine Learning Engineer',
       'AI Developer', 'Research Scientist', 'Data Analytics Manager',
       'Business Data Analyst', 'Applied Data Scientist',
       'Staff Data Analyst', 'ETL Engineer', 'Data DevOps Engineer',
       'Head of Data', 'Data Science Manager', 'Data Manager',
       'Machine Learning Researcher', 'Big Data Engineer',
       'Data Specialist', 'Lead Data Analyst', 'BI Data Engineer',
       'Director of Data Science', 'Machine Learning Scientist',
       'MLOps Engineer', 'AI Scientist', 'Autonomous Vehicle Technician',
       'Applied Machine Learning Sc

In [91]:
df.shape

(3755, 11)

In [92]:
top_titles = df['job_title'].value_counts().nlargest(20).index

df['job_title_grouped'] = df['job_title'].apply(
    lambda x: x if x in top_titles else 'Other'
)

In [93]:
df.drop(
    columns=["job_title", "salary_currency", "salary"],
    inplace=True
)

print(df.columns)

Index(['work_year', 'experience_level', 'employment_type', 'salary_in_usd',
       'employee_residence', 'remote_ratio', 'company_location',
       'company_size', 'job_title_grouped'],
      dtype='object')


In [94]:
df = df.drop("company_location",axis=1)

In [95]:
df.shape

(3755, 8)

In [96]:
df.head()

,work_year,experience_level,employment_type,salary_in_usd,employee_residence,remote_ratio,company_size,job_title_grouped
0,2023,SE,FT,85847,ES,100,L,Other
1,2023,MI,CT,30000,US,100,S,ML Engineer
2,2023,MI,CT,25500,US,100,S,ML Engineer
3,2023,SE,FT,175000,CA,100,M,Data Scientist
4,2023,SE,FT,120000,CA,100,M,Data Scientist


In [97]:

print(df.columns.tolist())

['work_year', 'experience_level', 'employment_type', 'salary_in_usd', 'employee_residence', 'remote_ratio', 'company_size', 'job_title_grouped']


In [98]:
df = pd.get_dummies(df,columns=["experience_level","employment_type","job_title_grouped","company_size"],drop_first=True)

In [100]:
df = df.drop("employee_residence",axis=1)

In [101]:
df.head()

,work_year,salary_in_usd,remote_ratio,experience_level_EX,experience_level_MI,experience_level_SE,employment_type_FL,employment_type_FT,employment_type_PT,job_title_grouped_Analytics Engineer,...,job_title_grouped_Data Scientist,job_title_grouped_Data Specialist,job_title_grouped_ML Engineer,job_title_grouped_Machine Learning Engineer,job_title_grouped_Machine Learning Scientist,job_title_grouped_Other,job_title_grouped_Research Engineer,job_title_grouped_Research Scientist,company_size_M,company_size_S
0,2023,85847,100,False,False,True,False,True,False,False,...,False,False,False,False,False,True,False,False,False,False
1,2023,30000,100,False,True,False,False,False,False,False,...,False,False,True,False,False,False,False,False,False,True
2,2023,25500,100,False,True,False,False,False,False,False,...,False,False,True,False,False,False,False,False,False,True
3,2023,175000,100,False,False,True,False,True,False,False,...,True,False,False,False,False,False,False,False,True,False
4,2023,120000,100,False,False,True,False,True,False,False,...,True,False,False,False,False,False,False,False,True,False


In [108]:
x = df.drop("salary_in_usd",axis=1)

In [109]:
y = df["salary_in_usd"]

In [123]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score



In [124]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [127]:
models={
    "LinearRegression": LinearRegression(),
    "RandomForest": RandomForestRegressor()
}

In [128]:
for i,model in models.items():
  model.fit(x_train,y_train)
  pred = model.predict(x_test)
  print("mean_absolute_error",mean_absolute_error(y_test,pred))
  print("mean_squared_error",mean_squared_error(y_test,pred))
  print("r2_score",r2_score(y_test,pred))



mean_absolute_error 42012.81703452802
mean_squared_error 2897192933.125654
r2_score 0.26612280443185055
mean_absolute_error 41606.41467442371
mean_squared_error 2975042420.040084
r2_score 0.24640303966232102


In [102]:
df.head()

,work_year,salary_in_usd,remote_ratio,experience_level_EX,experience_level_MI,experience_level_SE,employment_type_FL,employment_type_FT,employment_type_PT,job_title_grouped_Analytics Engineer,...,job_title_grouped_Data Scientist,job_title_grouped_Data Specialist,job_title_grouped_ML Engineer,job_title_grouped_Machine Learning Engineer,job_title_grouped_Machine Learning Scientist,job_title_grouped_Other,job_title_grouped_Research Engineer,job_title_grouped_Research Scientist,company_size_M,company_size_S
0,2023,85847,100,False,False,True,False,True,False,False,...,False,False,False,False,False,True,False,False,False,False
1,2023,30000,100,False,True,False,False,False,False,False,...,False,False,True,False,False,False,False,False,False,True
2,2023,25500,100,False,True,False,False,False,False,False,...,False,False,True,False,False,False,False,False,False,True
3,2023,175000,100,False,False,True,False,True,False,False,...,True,False,False,False,False,False,False,False,True,False
4,2023,120000,100,False,False,True,False,True,False,False,...,True,False,False,False,False,False,False,False,True,False


In [103]:
print(type(df))
print(df.head())
print(df.shape)

<class 'pandas.core.frame.DataFrame'>
   work_year  salary_in_usd  remote_ratio  experience_level_EX  \
0       2023          85847           100                False   
1       2023          30000           100                False   
2       2023          25500           100                False   
3       2023         175000           100                False   
4       2023         120000           100                False   

   experience_level_MI  experience_level_SE  employment_type_FL  \
0                False                 True               False   
1                 True                False               False   
2                 True                False               False   
3                False                 True               False   
4                False                 True               False   

   employment_type_FT  employment_type_PT  \
0                True               False   
1               False               False   
2               False      

In [129]:
from sklearn.dummy import DummyRegressor
from sklearn.metrics import r2_score

dummy = DummyRegressor(strategy="mean")
dummy.fit(x_train, y_train)

pred = dummy.predict(x_test)

print(r2_score(y_test, pred))

-0.0014932839167356438


In [130]:
importance = pd.DataFrame({
    'feature': x_train.columns,
    'importance': model.feature_importances_
})

importance = importance.sort_values(
    by='importance',
    ascending=False
)

print(importance.head(10))

                                 feature  importance
4                    experience_level_SE    0.248376
2                    experience_level_EX    0.143913
0                              work_year    0.104165
1                           remote_ratio    0.092987
13        job_title_grouped_Data Analyst    0.075997
28                        company_size_M    0.044745
25               job_title_grouped_Other    0.030505
29                        company_size_S    0.027527
27  job_title_grouped_Research Scientist    0.026887
16       job_title_grouped_Data Engineer    0.023278


In [136]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("arnabchaki/data-science-salaries-2023")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'data-science-salaries-2023' dataset.
Path to dataset files: /kaggle/input/data-science-salaries-2023


In [138]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import os


# Load data

df = pd.read_csv(os.path.join(path,"ds_salaries.csv"))

# -----------------------------
# Job Title Grouping
# -----------------------------
job_mapping = {
    "Data Scientist": "Data Scientist",
    "Senior Data Scientist": "Data Scientist",
    "Lead Data Scientist": "Data Scientist",

    "Data Analyst": "Data Analyst",
    "Business Data Analyst": "Data Analyst",

    "Data Engineer": "Data Engineer",
    "Senior Data Engineer": "Data Engineer",

    "Machine Learning Engineer": "ML Engineer",
    "ML Engineer": "ML Engineer",

    "Research Scientist": "Research Scientist"
}

df["job_title_grouped"] = df["job_title"].map(job_mapping)
df["job_title_grouped"] = df["job_title_grouped"].fillna("Other")

# -----------------------------
# Drop unnecessary columns
# -----------------------------
df.drop(
    columns=[
        "job_title",
        "salary",
        "salary_currency"
    ],
    inplace=True
)

# -----------------------------
# One Hot Encoding
# -----------------------------
df = pd.get_dummies(
    df,
    columns=[
        "experience_level",
        "employment_type",
        "job_title_grouped",
        "employee_residence",
        "company_location",
        "company_size"
    ],
    drop_first=True
)

# -----------------------------
# Features and Target
# -----------------------------
X = df.drop("salary_in_usd", axis=1)
y = df["salary_in_usd"]

# -----------------------------
# Train Test Split
# -----------------------------
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# -----------------------------
# Models
# -----------------------------
models = {
    "Linear Regression": LinearRegression(),
    "Random Forest": RandomForestRegressor(
        n_estimators=200,
        random_state=42
    )
}

# -----------------------------
# Training and Evaluation
# -----------------------------
for name, model in models.items():

    model.fit(X_train, y_train)

    pred = model.predict(X_test)

    print("\n", "="*50)
    print(name)
    print("="*50)

    print("MAE :", mean_absolute_error(y_test, pred))
    print("MSE :", mean_squared_error(y_test, pred))
    print("R2  :", r2_score(y_test, pred))

# -----------------------------
# Random Forest Feature Importance
# -----------------------------
rf = models["Random Forest"]

importance = pd.DataFrame({
    "feature": X_train.columns,
    "importance": rf.feature_importances_
})

importance = importance.sort_values(
    by="importance",
    ascending=False
)

print("\nTop 20 Features")
print(importance.head(20))


Linear Regression
MAE : 37794.24224907525
MSE : 2415369685.5883417
R2  : 0.38817166407777626

Random Forest
MAE : 37084.755018808006
MSE : 2415305613.4964366
R2  : 0.3881878939417195

Top 20 Features
                                  feature  importance
87                  employee_residence_US    0.406580
4                     experience_level_SE    0.082507
0                               work_year    0.072581
2                     experience_level_EX    0.055446
1                            remote_ratio    0.048077
8         job_title_grouped_Data Engineer    0.036274
12   job_title_grouped_Research Scientist    0.035075
11                job_title_grouped_Other    0.030115
10          job_title_grouped_ML Engineer    0.025755
9        job_title_grouped_Data Scientist    0.021713
161                        company_size_M    0.021136
3                     experience_level_MI    0.020439
23                  employee_residence_CA    0.017882
162                        company_size_S  

## In the previous model, company_location and employee_residence were dropped, causing the model to miss important geographic salary patterns.
# After including location features, employee_residence_US emerged as the most influential predictor, confirming the importance of geography in salary prediction.
# Model performance improved significantly, with R² increasing from ~0.26 to ~0.39, while experience level, work year, and job role remained key factors.